solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttpSolrClientBase.java

```diff
@@ -0,0 +1,288 @@
+import org.apache.solr.client.solrj.request.RequestWriter;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.util.NamedList;
+import org.slf4j.MDC;
+
+/**
+ * LBHttp2SolrClient or "LoadBalanced LBHttp2SolrClient" is a load balancing wrapper around {@link
+ * Http2SolrClient}. This is useful when you have multiple Solr endpoints and requests need to be
+ * Load Balanced among them.
+ *
+ * <p>Do <b>NOT</b> use this class for indexing in leader/follower scenarios since documents must be
+ * sent to the correct leader; no inter-node routing is done.
+ *
+ * <p>In SolrCloud (leader/replica) scenarios, it is usually better to use {@link CloudSolrClient},
+ * but this class may be used for updates because the server will forward them to the appropriate
+ * leader.
+ *
+ * <p>It offers automatic failover when a server goes down, and it detects when the server comes
+ * back up.
+ *
+ * <p>Load balancing is done using a simple round-robin on the list of endpoints. Endpoint URLs are
+ * expected to point to the Solr "root" path (i.e. "/solr").
+ *
+ * <blockquote>
+ *
+ * <pre>
+ * SolrClient client = new LBHttp2SolrClient.Builder(http2SolrClient,
+ *         new LBSolrClient.Endpoint("http://host1:8080/solr"), new LBSolrClient.Endpoint("http://host2:8080/solr"))
+ *     .build();
+ * </pre>
+ *
+ * </blockquote>
+ *
+ * Users who wish to balance traffic across a specific set of replicas or cores may specify each
+ * endpoint as a root-URL and core-name pair. For example:
+ *
+ * <blockquote>
+ *
+ * <pre>
+ * SolrClient client = new LBHttp2SolrClient.Builder(http2SolrClient,
+ *         new LBSolrClient.Endpoint("http://host1:8080/solr", "coreA"),
+ *         new LBSolrClient.Endpoint("http://host2:8080/solr", "coreB"))
+ *     .build();
+ * </pre>
+ *
+ * </blockquote>
+ *
+ * <p>If a request to an endpoint fails by an IOException due to a connection timeout or read
+ * timeout then the host is taken off the list of live endpoints and moved to a 'dead endpoint list'
+ * and the request is resent to the next live endpoint. This process is continued till it tries all
+ * the live endpoints. If at least one endpoint is alive, the request succeeds, and if not it fails.
+ *
+ * <p>Dead endpoints are periodically healthchecked on a fixed interval controlled by {@link
+ * LBHttp2SolrClient.Builder#setAliveCheckInterval(int, TimeUnit)}. The default is set to one
+ * minute.
+ *
+ * <p><b>When to use this?</b><br>
+ * This can be used as a software load balancer when you do not wish to set up an external load
+ * balancer. Alternatives to this code are to use a dedicated hardware load balancer or using Apache
+ * httpd with mod_proxy_balancer as a load balancer. See <a
+ * href="http://en.wikipedia.org/wiki/Load_balancing_(computing)">Load balancing on Wikipedia</a>
+ *
+ * @since solr 8.0
+ */
+public abstract class LBHttpSolrClientBase<C extends HttpSolrClientBase> extends LBSolrClient {
```

Comment:
Do we need subclassing?  In other words, can we have a general (not abstract) LB SolrClient that accepts an HttpSolrClientBase ?

Reply:
Yes, we do not need to subclass and I agree a generic LBHttpSolrClient would be better than this.  Doing this would be an API change, and although we can change APIs with 10.0 coming, I wonder if it is worth annoying our users for what seems like a very slight improvement?

Reply:
I think yes; if we can't change major things in a major API then.... ? wow.
Also, again, few users use LbSolrClient; it's more of an internal SolrJ component.  I say this from experience only; we don't officially communicate such.